### PR TITLE
Tracing

### PR DESCRIPTION
**Pull request by JoelEinbinder**
#179

### CODE REVIEW

File: docs/api.md [hunk 1]
```diff
@@ -827,6 +834,26 @@ Dispatches a `mousemove` event.
...
+
+#### tracing.start([options])
+- `options` <[Object]>
+  - `screenshots` <[boolean]> captures screenshots in the trace.
+  - `clickCount` <[number]> defaults to 1
+- returns: <[Promise]>
+
+#### tracing.stop(path)
+- `path` <[string]> A path to write the trace file to.
+- returns: <[Promise]>
```
Comment (aslushnikov) [inline docs/api.md:854]: - should return <[Promise<[string]>>
- path is optional

File: lib/Tracing.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+  /**
+   * @param {Object=} options
+   */
+  async start(options = {}) {
+    if (this._recording)
+      throw new Error('Cannot start recording trace while already recording trace.');
+
+    const categoriesArray = [
+      '-*', 'devtools.timeline', 'v8.execute', 'disabled-by-default-devtools.timeline',
```
Comment (aslushnikov) [inline lib/Tracing.js]: Where do these categories come from?

File: docs/api.md [hunk 3]
```diff
@@ -827,6 +834,26 @@ Dispatches a `mousemove` event.
...
+```js
+await page.tracing.start();
+await page.navigate('https://www.google.com');
+await page.tracing.stop('trace.json');
+```
+
+#### tracing.start([options])
+- `options` <[Object]>
+  - `screenshots` <[boolean]> captures screenshots in the trace.
+  - `clickCount` <[number]> defaults to 1
```
Comment (aslushnikov) [inline docs/api.md]: what is click count??

File: lib/Tracing.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+      transferMode: 'ReturnAsStream',
+      categories: categoriesArray.join(',')
+    });
+  }
+
+  /**
+   * @param {string} path
+   */
+  async stop(path) {
+    this._recording = false;
```
Comment (aslushnikov) [inline lib/Tracing.js]: let's have it after the trace is fetched from backend

File: lib/Tracing.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    let fulfill;
+    let contentPromise = new Promise(x => fulfill = x);
+    this._client.once('Tracing.tracingComplete', event => {
+      this._readStream(event.stream, path).then(fulfill);
+    });
+    await this._client.send('Tracing.end', {
```
Comment (aslushnikov) [inline lib/Tracing.js]: let's remove empty multiline arguments object

File: docs/api.md (hunk 3 above)
Comment (JoelEinbinder) [reply docs/api.md]: That keeps getting added. I've deleted it 3 times..

File: docs/api.md (hunk 1 above)
Comment (JoelEinbinder) [reply docs/api.md:854]: Talking with @paulirish , large strings might kill node. So we force it to be written to disk.

File: lib/Tracing.js (hunk 2 above)
Comment (JoelEinbinder) [reply lib/Tracing.js]: Default devtools trace.

**Comment by JoelEinbinder** [comment]
I enabled JS profiling, and fixed the other concerns.

File: test/test.js [hunk 6]
```diff
@@ -1481,6 +1481,19 @@ describe('Page', function() {
...
+    let outputFile = path.join(__dirname, 'assets', 'trace.json');
+    afterEach(function() {
+      fs.unlinkSync(outputFile);
+    });
+    it('should output a trace', SX(async function() {
+      await page.tracing.start({screenshots: true});
+      await page.navigate(PREFIX + '/grid.html');
+      await page.tracing.stop(outputFile);
+      expect(fs.existsSync(outputFile)).toBe(true);
+    }));
```
Comment (aslushnikov) [inline test/test.js:1495]: let's add a test which starts tracing in parallel in two pages and makes sure we fail

File: lib/Tracing.js [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+  constructor(client) {
+    this._client = client;
+    this._recording = false;
+  }
+
+  /**
+   * @param {Object=} options
+   */
+  async start(options = {}) {
+    if (this._recording)
```
Comment (aslushnikov) [inline lib/Tracing.js]:
```js
console.assert(!this._recording, 'Cannot start recording trace while already recording trace.')
```

File: lib/Tracing.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+      let response = await this._client.send('IO.read', {handle});
+      eof = response.eof;
+      if (path)
+        fs.writeSync(file, response.data);
+    }
+    fs.closeSync(file);
+    await this._client.send('IO.close', {handle});
+  }
+}
+module.exports = Tracing;
```
Comment (aslushnikov) [inline lib/Tracing.js]: this should also have the "tracePublicAPI" call to helper

File: test/test.js (hunk 6 above)
Comment (JoelEinbinder) [reply test/test.js:1495]: Done.

File: lib/Tracing.js (hunk 7 above)
Comment (JoelEinbinder) [reply lib/Tracing.js]: Done.

File: lib/Page.js [hunk 9]
```diff
@@ -23,6 +23,7 @@ let Dialog = require('./Dialog');
 let EmulationManager = require('./EmulationManager');
 let FrameManager = require('./FrameManager');
 let {Keyboard, Mouse} = require('./Input');
+let Tracing = require('./Tracing');
```
Comment (a1ph) [inline lib/Page.js:26]: const?

File: lib/Tracing.js (hunk 8 above)
Comment (JoelEinbinder) [reply lib/Tracing.js]: Done.

File: test/test.js [hunk 10]
```diff
@@ -1481,6 +1481,31 @@ describe('Page', function() {
...
+      await page.tracing.start();
+      let newPage = await browser.newPage();
+      let error = null;
+      try {
+        await newPage.tracing.start();
+      } catch (e) {
+        error = e;
+      }
+      expect(error).toBeTruthy();
+      await page.tracing.stop(outputFile);
```
Comment (aslushnikov) [inline test/test.js]: Let's close the page so that it's not dangling for the rest of the tests
```js
await newPage.close();
```

**Review by aslushnikov — APPROVED** [review]

File: lib/Tracing.js [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+      categories: categoriesArray.join(',')
+    });
+  }
+
+  /**
+   * @param {string} path
+   */
+  async stop(path) {
+    let fulfill;
+    let contentPromise = new Promise(x => fulfill = x);
```
Comment (a1ph) [inline lib/Tracing.js:56]: Why all this promise juggling? Sequential code would look more clear, e.g.:
```js
this._client.send('Tracing.end');
const {stream} = await this._client.once('Tracing.tracingComplete');
await this._readStream(stream, path);
this._recording = false;
```

File: lib/Tracing.js [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+
+    this._recording = true;
+    await this._client.send('Tracing.start', {
+      transferMode: 'ReturnAsStream',
+      categories: categoriesArray.join(',')
+    });
+  }
+
+  /**
+   * @param {string} path
```
Comment (a1ph) [inline lib/Tracing.js:52]: @return {!Promise}

File: lib/Tracing.js [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+  }
+
+  /**
+   * @param {string} handle
+   * @param {string=} path
+   * @return {!Promise}
+   */
+  async _readStream(handle, path) {
+    let eof = false;
+    let file = fs.openSync(path, 'w');
```
Comment (a1ph) [inline lib/Tracing.js:72]: const
Also what if path is undefined?

File: lib/Tracing.js [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+    while (!eof) {
+      let response = await this._client.send('IO.read', {handle});
```
Comment (a1ph) [inline lib/Tracing.js:74]: const

File: test/test.js (hunk 10 above)
Comment (JoelEinbinder) [reply test/test.js]: Done.

File: test/test.js [hunk 15]
```diff
@@ -1481,6 +1481,19 @@ describe('Page', function() {
       await Promise.all(pages.map(page => page.close()));
     }));
   });
+
+  describe('Tracing', function() {
+    let outputFile = path.join(__dirname, 'assets', 'trace.json');
```
Comment (a1ph) [inline test/test.js:1486]: const

File: lib/Page.js (hunk 9 above)
Comment (aslushnikov) [reply lib/Page.js:26]: We need to convert all our `require`'s into consts, let's do it separately

File: lib/Tracing.js (hunk 12 above)
Comment (aslushnikov) [reply lib/Tracing.js:52]: Let's address this in a follow-up (I've already merged the PR before I saw your comments)

File: lib/Tracing.js (hunk 11 above)
Comment (aslushnikov) [reply lib/Tracing.js:56]: `once` doesn't return a promise

File: lib/Tracing.js (hunk 13 above)
Comment (aslushnikov) [reply lib/Tracing.js:72]: path is not optional in `trace.stop()` method

File: docs/api.md [hunk 16]
```diff
@@ -827,6 +834,25 @@ Dispatches a `mousemove` event.
...
+await page.tracing.stop('trace.json');
+```
+
+#### tracing.start([options])
+- `options` <[Object]>
+  - `screenshots` <[boolean]> captures screenshots in the trace.
+- returns: <[Promise]>
+
+#### tracing.stop(path)
+- `path` <[string]> A path to write the trace file to.
```
Comment (paulirish) [inline docs/api.md:853]: not options object?  
screenshot takes path inside the object, so this feels a tad inconsistent

Comment (paulirish) [reply docs/api.md:853]: actually... it makes a bit more sense to specify this path on start() instead.  yah?

**Pull request merged**